lib/cretonne/meta/cdsl/ast.py

```diff
@@ -332,14 +368,51 @@ def inst_predicate(self):
         return pred
 
     def copy(self, m):
-        # type: (Dict[Var, Var]) -> Apply
+        # type: (VarMap) -> Apply
         """
         Return a copy of this Expr with vars replaced with fresh variables,
         in accordance with the map m. Update m as neccessary.
         """
         return Apply(self.inst, tuple(map(lambda e: replace_var(e, m),
                                           self.args)))
 
+    def vars(self):
+        # type: () -> Set[Var]
+        """ Return the set of all Vars that appear in self"""
+        res = set()
+        for i in self.inst.value_opnums:
+            arg = self.args[i]
+            assert isinstance(arg, Var)
+            res.add(arg)
+        return res
+
+    def substitution(self, other, s):
+        # type: (Apply, VarMap) -> Optional[VarMap]
+        """
+        If the application self and other agree structurally, return a variable
+        substitution to transform self ot other. Two applications agree
```

Comment:
typo: ot
add: "Otherwise, return `None`".

That other mention of "agree structurally" can simply refer to the definition here.